gnucash/gnome-utils/gnc-autoclear.cpp

```diff
@@ -0,0 +1,280 @@
+        : reachable_amount (amount), splits_vector(splits){}
+};
+
+static void status_update (GtkLabel *label, gchar *status)
+{
+    if (!label) return;
+    gtk_label_set_text (label, status);
+    while (gtk_events_pending ())
+        g_main_context_iteration (nullptr, FALSE);
+}
+
+static void looping_update_status (GtkLabel *label, guint nc_progress,
+                                   guint nc_size, guint size,
+                                   clock_t *next_update_tick)
+{
+    clock_t now;
+    if (!label)
+        return;
+    now = clock ();
+    if (G_UNLIKELY (now > *next_update_tick))
+    {
+        gchar *text = g_strdup_printf ("%u/%u splits processed, %u combos",
+                                       nc_progress, nc_size, size);
+        status_update (label, text);
+        g_free (text);
+        *next_update_tick = now + MAX_UPDATE_TICKS;
+    }
+}
+
+class Hasher
+{
+public:
+    size_t operator() (gint64 const& key) const
+    {
+        return key;
+    }
+};
+class EqualFn
+{
+public:
+    bool operator() (gint64 const& t1, gint64 const& t2) const
+    {
+        return (t1 == t2);
+    }
+};
+
+gboolean
+gnc_autoclear_get_splits (Account *account, gnc_numeric toclear_value,
+                          time64 end_date,
+                          GList **splits, GError **error, GtkLabel *label)
+{
+    std::vector<std::shared_ptr<Split>> nc_vector, DUP_VEC;
+    std::unordered_map<gint64, std::vector<std::shared_ptr<Split>>,
+                       Hasher, EqualFn> sack;
+    guint nc_progress = 0;
+    clock_t start_ticks, next_update_tick;
+    gboolean debugging_enabled = qof_log_check (G_LOG_DOMAIN, QOF_LOG_DEBUG);
+    GQuark autoclear_quark = g_quark_from_static_string ("autoclear");
+
+    g_return_val_if_fail (GNC_IS_ACCOUNT (account), FALSE);
+    g_return_val_if_fail (splits != nullptr, FALSE);
+
+    DUP_VEC = {};
+    *splits = nullptr;
+
+    /* Extract which splits are not cleared and compute the amount we have to clear */
+    for (GList *node = xaccAccountGetSplitList (account); node; node = node->next)
+    {
+        auto split = static_cast<Split*> (node->data);
+
+        if (xaccSplitGetReconcile (split) != NREC)
+            toclear_value = gnc_numeric_sub_fixed
+                (toclear_value, xaccSplitGetAmount (split));
+        else if (gnc_numeric_zero_p (xaccSplitGetAmount (split)))
+            DEBUG ("skipping zero-amount split %p", split);
+        else if (gnc_numeric_check (xaccSplitGetAmount (split)))
+            DEBUG ("skipping invalid-amount split %p", split);
+        else if (end_date != INT64_MAX &&
+                 xaccTransGetDate (xaccSplitGetParent (split)) > end_date)
+            DEBUG ("skipping split after statement_date %p", split);
+        else
+        {
+            auto s = static_cast<std::shared_ptr<Split>>(split);
+            auto new_s = new Split;
+            memcpy (new_s, &s, sizeof(Split));
+            nc_vector.emplace_back (s);
```

Comment:
[emplace_back](https://en.cppreference.com/w/cpp/container/vector/emplace_back) will copy-construct a new shared_ptr<Split> from s; Since s's [control block ptr](https://www.nextptr.com/tutorial/ta1358374985/shared_ptr-basics-and-internals-with-examples) is undefined so will the one in the vector. If you wanted and had made a valid shared_ptr you'd normally do `nc_vector.push_back(std::move(s))` to quickly transfer s into the vector without the expense of copying and deleting it with the consequent two changes to the control block.